Generate Fio Job Files Automatically

noahl opened the pull request:
Make the fio benchmark generate job files automatically based on
user-specified parameters. Can still be overridden with a
user-supplied job file.

noahl commented:
@yuyantingzero @stfeng @cmccoy Would you take a look?

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -53,6 +57,96 @@
+  except ValueError:
+    pass
+
+  bounds = io_depths.split('-', 1)
```
Move this part inside except ValueError

noahl replied:
Done.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -61,53 +155,37 @@ def GetInfo():
+
+  if (FLAGS.fio_jobfile and
+      (FLAGS.against_device or FLAGS.device_fill_size or FLAGS.io_depths)):
+    logging.warning('Fio job file specified. Ignoring other fio options.')
```
Give more detailed logging here, current message is a bit confusing.
eg, job_file= ..., against_device=...,

noahl replied:
I added logging that tells you which options are being ignored. Does this satisfy your concern?

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -61,53 +155,37 @@ def GetInfo():
-  vm.PushFile(file_path)
+
+  mount_point = vm.scratch_disks[0].mount_point
+  logging.info('Umount scratch disk on %s at %s', vm, mount_point)
```
We only umount the device when testing against raw device. If you are creating a fio test file at mount point (/scratch0). The device must be mounted.

noahl replied:
Fixed. Thanks!

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -20,20 +20,21 @@
 from perfkitbenchmarker.packages import fio
 
+LOCAL_JOB_FILE_NAME = 'fio.job'  # used with vm_util.PrependTempDir()
+REMOTE_JOB_FILE_PATH = '~/fio.job'
```
Prefer something in `vm_util.VM_TMP_DIR`

noahl replied:
Done.

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -53,6 +57,95 @@
+rw={rwkind}
+iodepth={iodepth}
+
+"""
```
We've been using Jinja2 templates for dynamically-generated config files. Something like:

```
[global]
ioengine=libaio
invalidate=1
blocksize=4k
direct=1
runtime=10m
time_based
filename={{filename}}
do_verify=0
verify_fatal=0
randrepeat=0
size={{size}}

{% for rwkind in ('randread', 'randwrite') %}  # or an input variable
{% for iodepth in iodepths %}
[{{rwkind}}-io-depth-{{iodepth}}]
stonewall
rw={{rwkind}}
iodepth={{iodepth}}
{% endfor %}
{% endfor %}
```

I think that communicates the intent of the configuration slightly better, since the template text and composition aren't separated.

noahl replied:
Done. Thank you!

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -53,6 +57,95 @@
+  """
+
+  if FLAGS.fio_jobfile:
+    return open(FLAGS.fio_jobfile, 'r').read()
```
Prefer using a context manager here:
```{python}
with open(FLAGS.fio_jobfile, 'r') as fp:
  return fp.read()
```
which guarantees that the file will be closed.

noahl replied:
Done.

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -53,6 +57,95 @@
+    filename = mount_point
+    size = FLAGS.device_fill_size
+  else:
+    filename = os.path.join(mount_point, 'fio-temp-file')
```
Introduce a constant for `fio-temp-file`

noahl replied:
Done.

cmccoy replied:
Also, please use `posixpath.join` to create remote file paths to maintain support for a windows controller.

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -152,8 +221,10 @@ def Cleanup(benchmark_spec):
+  if not FLAGS.against_device and not FLAGS.fio_jobfile:
+    # If the user supplies their own job file, then they have to clean
+    # up after themselves, because we don't know their temp file name.
+    vm.RemoveFile(vm.GetScratchDir() + 'fio-temp-file')
```
`posixpath.join(vm.GetScratchDir(), 'fio-temp-file')`

yuyantingzero replied:
I am confused here, why we don't know their temp file name. PKB upload the job file to remote vm from data/ directory?

noahl replied:
Done. Also changed other uses of os.path to posixpath.

noahl replied:
We don't know their temp file name because we would have to parse their job file in order to know it. It would add a lot of complexity to handle all of the cases correctly, so I think it's best to just say that if the user supplies a file, they can handle cleanup. It shouldn't be too big a deal, because in the common case we just destroy the scratch disk after testing and the file goes away.

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -53,6 +57,95 @@
+    An iterable of integers.
+
+  Raises:
+    ValueError if FLAGS.io_depths doesn't follow a format it recognizes.
```
Consider [registering a validator](https://github.com/google/python-gflags/blob/master/gflags.py#L166).

noahl replied:
Done.

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -53,6 +57,95 @@
+  """
+
+  try:
+    return [int(io_depths)]
```
Consider checking that all io_depths are >= 0

noahl replied:
Done in the new validator.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -53,6 +57,95 @@
+                     for i in GetIODepths(FLAGS.io_depths))))
+
+
+def JobFileString(vm):
```
Method name seems confusing. Consider GetJobFileContent?

noahl replied:
I changed it to GenerateJobFileString, so it matches the name JobFileString below it.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -123,26 +200,18 @@ def Run(benchmark_spec):
-  fio_command = 'sudo %s --output-format=json %s' % (fio.FIO_PATH, fio_command)
+
+  fio_command = 'sudo %s --output-format=json %s' % (fio.FIO_PATH,
+                                                     REMOTE_JOB_FILE_PATH)
```
I am concerned about the REMOTE_JOB_FILE_PATH. When you push the path to vm, the user is perfkit. When you ran fio command, you are root(sudo). The home directory might be different, so please use absolute path in REMOTE_JOB_FILE_PATH.

noahl replied:
I changed REMOTE_JOB_FILE_PATH to be in vm_util.VM_TMP_DIR, which is defined as an absolute path.

noahl commented:
I've uploaded a new version that I think addresses all of the comments above. I also wanted to make sure everyone knows that this PR removes some functionality from the fio benchmark. Specifically, I removed the memory_multiple option, because it was very closely tied to the default fio job file, and I don't think that made sense in a world where we're generating lots of different files. I also removed the ability to scan user-supplied job files and change the filename parameter. I did that because I think that scanning for filename correctly in all cases is hard (there might be multiple filenames, or there might be no filename parameter), and if we can't do it in all cases, then it seems better to just say to the user that we can't rather than trying and maybe silently failing.

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -54,6 +76,93 @@
+    filename = posixpath.join(mount_point, DEFAULT_TEMP_FILE_NAME)
+    size = '100G'
+
+  return jinja2.Template(JOB_FILE_TEMPLATE).render(
```
Nit: no need to change, but [`VirtualMachine.RenderTemplate(...)`](https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/blob/master/perfkitbenchmarker/virtual_machine.py#L334) is similar.

If you leave as-is, could you set `undefined=jinja2.StrictUndefined` in the Template constructor? Otherwise missing variables will be left blank, which can be hard to debug.

noahl replied:
Done. I left it as jinja2.Template rather than VirtualMachine.RenderTemplate because that requires a path on the remote machine to render to, but we need the results on the local machine.

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -54,6 +76,93 @@
+    return range(int(match.group(1)), int(match.group(3)) + 1)
+
+
+def WriteJobFile(mount_point):
```
To the extent you can, prefer taking arguments to accessing the FLAGS global - it makes testing easier.

The naming is also a bit confusing...this doesn't actually write the fio job.

noahl replied:
Changed to take arguments rather than using FLAGS. This function also has a different name now.

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -54,6 +76,93 @@
+    size = FLAGS.device_fill_size
+  else:
+    filename = posixpath.join(mount_point, DEFAULT_TEMP_FILE_NAME)
+    size = '100G'
```
Could you make this a constant?
Will this work for disks < 100G?

yuyantingzero replied:
Why use different size?

noahl replied:
I've changed it so the size is the maximum of 100G or the size of the scratch disk. The maximum size is a constant. The size shouldn't actually affect our runs at all - since they're time-based, they will keep repeating their workload until they hit their time limit. I believe the only thing this affects is the size of the file we create when we're not running against a raw disk.

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -61,53 +170,47 @@ def GetInfo():
+
+  if FLAGS.fio_jobfile:
+    ignored_flags = []
+    if FLAGS.against_device:
```
Could you:

* Check `FLAGS[flag_name].present` instead of repeating the default values here?
* Convert these individual checks to a for loop?

noahl replied:
Done. I also have a new constant at the top of the file listing all of the arguments that will be ignored if the user supplies a job file.

cmccoy commented on tests/benchmarks/fio_benchmark_test.py:
```diff
@@ -0,0 +1,31 @@
+    self.assertEqual(list(fio_benchmark.GetIODepths('3')), [3])
+
+  def testIntegerRange(self):
+    self.assertEqual(list(fio_benchmark.GetIODepths('3-5')), [3, 4, 5])
```
Could you add a test for `WriteJobFile`? Even a sanity check that the template renders successfully would be helpful.

noahl replied:
Done.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -61,53 +170,47 @@ def GetInfo():
+    if ignored_flags:
+      logging.warning('Fio job file specified. Ignoring options "%s"',
+                      ', '.join(ignored_flags))
+  if FLAGS.device_fill_size != '100%' and not FLAGS.against_device:
```
This flag is actually used in Prepare stage if writing to raw device.
For example, to use AWS GP. They require prewarm, basically sequentially writting the whole device.

noahl replied:
I changed it this morning so we won't prewarm if the user supplies a job file. I think that makes sense to give the user the most flexibility. They can always put a prewarm job in their job file if they want to.

noahl commented:
I signed it!

noahl commented:
The latest revision should fix Connor's points above. The one thing I haven't responded to is Yanting's request for a metadata label that would go on user-specified jobfile runs (or maybe on automatically-generated-jobfile runs) to let us easily see which tests have had the disk prewarmed and which have not. The goal is to prevent accidental apples-to-oranges comparisons when making dashboards.

I want to check what other people think before implementing, but I think a metadata tag like 'generated-jobfile' would be best. We don't know if user jobfiles prewarm or not, so it seems unfair to label them 'not-prewarmed'. Does that work for everyone?

noahl commented:
A quick note that I ran the current code with disk sizes 50GB and 100GB and it seemed to work, which should mean that scaling the size of the IO down with the size of the disk works properly.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -61,35 +188,40 @@ def GetInfo():
-    vm.RemoteCommand('sudo umount %s' % vm.GetScratchDir())
+
+  disk = vm.scratch_disks[0]
+  if FLAGS.against_device and not FLAGS.fio_jobfile:
```
A bit confused here. The only chance we test against device, is when using the autogenerated job file?

noahl replied:
Yes, that's how it's set up. I realized after reading your comment that people using custom job files still probably need an option to test against the disk, because you can't unmount a disk from a fio jobfile. I can fix that.

yuyantingzero replied:
Cool. Thanks.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -54,6 +80,107 @@
                   'num_machines': 1}
 
 
+JOB_FILE_TEMPLATE = """
```
We need be careful if replacing the old job file with this new template. This seems to be a completely new benchmark.

noahl replied:
You're right, and the results are not comparable to the results from the old benchmark. We don't want it to be too easy to confuse the two benchmarks. I'm not sure how to handle this properly. I think the choices are either put something in the metadata to indicate what the benchmark is, make my code a new benchmark, or maybe have fio_benchmark record samples with multiple test names (I was thinking that each rw mode could be its own test, so "read", "randread", "write", and "randwrite" would all be different). I can argue for each of these options, but I can't convince myself that one of them is clearly right.

stfeng replied:
@yuyantingzero @noahl 

I am curious to know why using a job file template here makes it a "completely new benchmark". It's still FIO benchmark. :)

Perhaps it's because we are no longer testing directly against device if we use the template? but we can still specify the filename to something like '/dev/sdb' to test directly against device, right? 

Let me know if I am missing something here...

noahl replied:
@stfeng 

I agreed that it was a new benchmark because I thought the results from the new code wouldn't be comparable to results from the old code. I just checked the job files, and actually the old and new use the same options (I think I did this on purpose), so they are more comparable than I thought.

The new jobfile generation code will test directly against the device if you use the --against_device argument.

yuyantingzero replied:
@stfeng @noahl The default behavior of the new code is doing random RW with varying queue depth, while the default behavior of the old code is doing seq RW, ran RW, which is more like a overview of the disk performance.
If user is using their own fio.job, this PR won't make any difference. However, if someone write some automation using the default old fio benchmark, the published metric will definitely be different and we might break their automation.
I suggest, we should at least mention in CHANGES.md that this is a breaking change that changes the default behavior of the fio benchmark, or go through the community, which is far more complex, since this benchmark is in standard_set.

stfeng replied:
Sgtm. Yes these are comparable job files so the tests should be comparable
too.

Thanks for the explanation.

On Aug 20, 2015, at 9:42 PM, noahl <notifications@github.com> wrote:

In perfkitbenchmarker/benchmarks/fio_benchmark.py
<https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/421#discussion_r37606266>
:

> @@ -54,6 +80,107 @@
>                    'num_machines': 1}
>
>
> +JOB_FILE_TEMPLATE = """

@stfeng <https://github.com/stfeng>

I agreed that it was a new benchmark because I thought the results from the
new code wouldn't be comparable to results from the old code. I just
checked the job files, and actually the old and new use the same options (I
think I did this on purpose), so they are more comparable than I thought.

The new jobfile generation code will test directly against the device if
you use the --against_device argument.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/421/files#r37606266>
.

stfeng replied:
I see. Thanks yanting!

Noah, is it possible to not change the default behavior with your new code?
That is, by default, still use the old job files. And opt-in to use the new
template-based job files generation logic (doing random rw with varying
queue depths).



On Aug 20, 2015, at 10:03 PM, yuyanting <notifications@github.com> wrote:

In perfkitbenchmarker/benchmarks/fio_benchmark.py
<https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/421#discussion_r37606815>
:

> @@ -54,6 +80,107 @@
>                    'num_machines': 1}
>
>
> +JOB_FILE_TEMPLATE = """

@stfeng <https://github.com/stfeng> @noahl <https://github.com/noahl> The
default behavior of the new code is doing random RW with varying queue
depth, while the default behavior of the old code is doing seq RW, ran RW,
which is more like a overview of the disk performance.
If user is using their own fio.job, this PR won't make any difference.
However, if someone write some automation using the default old fio
benchmark, the published metric will definitely be different and we might
break their automation.
I suggest, we should at least mention in CHANGES.md that this is a breaking
change that changes the default behavior of the fio benchmark, or go
through the community, which is far more complex, since this benchmark is
in standard_set.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/421/files#r37606815>
.

noahl replied:
@stfeng @yuyantingzero 

I've been thinking about this, and I'd like to propose that we handle the change by generalizing the template-based code so it can generate something similar to the old fio.job file, and then making that the default. My vision for the new code is that it lets you essentially take the old job file and vary some parameters to see what happens. Right now it only lets you vary queue depth, but we could easily make it vary more parameters without changing the interface.

The parts of the old job file that will be difficult to generate exactly with my new code are the fact that the jobs use different sizes, the fact that the sizes depend on $mb_memory, and the fact that one job has end_fsync=1 and none of the others do. If you two think those are important things to preserve, then I can make it use the old job file by default. But I think it would be much cleaner to have the default be generating a test that happens to look like the old fio.job file, while giving the user the option to still run exactly the old fio.job file.

Also, since we're hoping to release 1.0 soon, I think we should be asking not just will this change break someone's automation, but also, what are the defaults that we want to keep for a long time? If we're going to change, I'd much rather do it before 1.0 than after it.

stfeng replied:
We should decouple changing the default of FIO benchmarks from this pr. This PR is about introducing a way to dynamically generate a job file via a template and use it for various scenarios. 

If it's easy to use the new way to generate the same job file as the test was using before as default, great. Otherwise, we should just retain those job files and focus this PR on its original purpose.

We can address the default behavior of this benchmark in a separate discussion/PR.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -20,24 +20,62 @@
-flags.DEFINE_string('fio_jobfile', 'fio.job', 'job file that fio will use')
-flags.DEFINE_integer('memory_multiple', 10,
-                     'size of fio scratch file compared to main memory size.')
+flags.DEFINE_string('fio_jobfile', None,
```
By default, we are still reading the fio.job, consider using "fio.job" as default value.

noahl replied:
This patch changes the semantics of the flag so the fio_jobfile is a path in the user's file system, not a path in our data directory. I think it is confusing now to make the default be 'fio.job' when we don't look for a file of that name. I will add a note in the help string saying what the default is.

yuyantingzero replied:
Thanks. Please also add a note saying it is looking for absolute directory.

noahl replied:
I don't think it should look for an absolute directory, actually. I'd like it to behave the same way as other command-line utilities, so the path is relative to your working directory. However, when you pass it a job file, it always copies it to its temporary directory.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -20,24 +20,62 @@
-                     'Test direct against scratch disk. If set True, will '
-                     'create a modified job file in temp directory, '
-                     'which ignores directory and filename parameter.')
+                     'If --generate_scenarios is given, test against the raw '
```
I think the documentation here is confusing. From the code, whenever against_device is True, the disk is umounted.

noahl replied:
I agree - thanks for catching it! I'll fix the documentation.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -54,6 +113,122 @@
+  if fio_jobfile:
+    with open(fio_jobfile, 'r') as jobfile:
+      return jobfile.read()
+  else:
```
else block will never be executed, right? Since you reset the fio_jobfile in line 214

yuyantingzero replied:
If scenario_strings is None, else block will never be executed.

noahl replied:
The else block can be executed if the user passes --generate_scenarios and does not pass --fio_jobfile. In that case, scenario_strings will either be [] or a list of strings and line 214 will not reset fio_jobfile, so it is possible to hit the else case. I will make the code more clear.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -61,53 +236,72 @@ def GetInfo():
+  else:
+    file_path = posixpath.join(disk.mount_point, DEFAULT_TEMP_FILE_NAME)
+    # We compute in MB to avoid rounding errors with 1GB disks.
+    size = str(
```
You can directly specify G in job file, for example size=100G. Also consider adding a flag for size.

noahl replied:
This size only applies to the prewarming, not the job file. I will make that more clear.

The regular size parameter doesn't apply to the generated files, because they use time-based runs, but I do plan to add a flag for working set size. You can expect that patch very soon.

yuyantingzero replied:
This works fine for the generated fio job. But if user provides a customized fio job with different size and different filename. Current warmup step won't writie the blocks we want to test. @stfeng , what is your suggestion about doing warming up with disk mounted?

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -61,53 +236,72 @@ def GetInfo():
+      device_path = disk.GetDevicePath()
+      logging.info('Fill scratch disk on %s at %s', vm, device_path)
+      command = (
+          ('sudo %s --filename=%s --ioengine=libaio '
```
Consider making command string a constant

noahl replied:
I like your idea a lot, but I ended up factoring the entire process of generating the command out to a global function. It feels like a function already.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -61,53 +264,69 @@ def GetInfo():
+  else:
+    file_path = posixpath.join(disk.mount_point, DEFAULT_TEMP_FILE_NAME)
+    # We compute in MB to avoid rounding errors with 1GB disks.
+    fill_size = str(
```
Sorry to make this complicated. To make warmup more reliable:
If fill_device=True, umount the disk, write the device. If not against_device, format and mount the device back.

noahl replied:
I tried this and ran into trouble. Eventually Eric helped me figure out why, and I think this will not work.

The reason is that when you fill the unmounted block device, you also get rid of the filesystem structure, so you can't re-mount the device. You could initialize a new filesystem, but then when you're reading from a large file, the file system will consider it unwritten, which I believe means it might just return zeros for untouched blocks without touching the block device. Because of this, I think we have to initialize through files. I've made the code for that cleaner and pushed a new revision.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -20,24 +20,61 @@
+flags.DEFINE_list('generate_scenarios', None,
+                  'Generate a job file with the given scenarios. Special '
+                  'scenario \'all\' generates all scenarios. Cannot use with '
+                  '--fio_jobfile.')
```
List available scenarios in help string.

noahl replied:
Done

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -54,6 +135,128 @@
+        rwkind=scenario['rwkind'],
+        blocksize=scenario['blocksize'],
+        iodepths=io_depths,
+        size=size_string))
```
If I remember it correctly, if you are testing against file system, size cannot be percentage.

noahl replied:
I just tested, and this appears to work.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -61,53 +264,68 @@ def GetInfo():
+    command = GenerateFillCommand(fio.FIO_PATH,
+                                  fill_path,
+                                  fill_size)
     vm.RemoteCommand(command)
```
The default behavior is changed to if we do nothing with the device_fill_size, we would fill the device? Please at least document the default behavior to avoid confusions.

noahl replied:
The default behavior is the same if you use --against_device, because the default value of --device_fill_size is '100%'. It's different if you don't use --against_device, because now we actually know how to fill file systems. I'll make sure the help text says what happens.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -54,6 +135,128 @@
+
+def GenerateJobFileString(disk, against_device,
+                          scenarios, io_depths, working_set_size):
+  """Write a fio job file.
```
The documentation for this method is not accurate. This method does not write a job file, it just generate the job file string.

noahl replied:
I changed it. Thank you for spotting this!

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -54,6 +135,128 @@
+
+def JobFileString(fio_jobfile, disk, against_device,
+                  scenario_strings, io_depths, working_set_size):
+  """Get the contents of our job file.
```
Be specific about the method. It does not simply get the contents, it also generate a new job file.

noahl replied:
Fixed. Thank you!

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -54,6 +135,128 @@
+  return file_string
+
+
+def JobFileString(fio_jobfile, disk, against_device,
```
Consider rename the method name to be something more specific.

noahl replied:
I've been thinking about this, but I don't know what to call it. Do you have any ideas?

yuyantingzero replied:
What about GetOrGenerateJobFileString? Just a suggestion, feel free to ignore.

noahl replied:
I used your suggestion. Thank you!

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -54,6 +135,128 @@
+      generate.
+    io_depths: iterable. The IO queue depths to test.
+    working_set_size: int or None. If int, the size of the working set in GB.
+    vm: the virtual_machine.BaseVirtualMachine that we will run on.
```
the vm doesn't appear in the parameter list of the method.

noahl replied:
Fixed. Thank you.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -123,26 +341,25 @@ def Run(benchmark_spec):
         If a 4th element is included, it is a dictionary of sample
```
Return section is outdated, it returns Sample.sample object.

noahl replied:
Fixed.

yuyantingzero commented:
A few comments about readability, otherwise LGTM

yuyantingzero commented on tests/benchmarks/fio_benchmark_test.py:
```diff
@@ -0,0 +1,59 @@
+
+  def testAgainstDevice(self):
+    # This just checks that the template renders.
+    fio_benchmark.GenerateJobFileString(
```
Could you verify the tests here by self.assertEqual(expected_job, generated_job)?

noahl replied:
Done.

yuyantingzero commented on tests/benchmarks/fio_benchmark_test.py:
```diff
@@ -0,0 +1,59 @@
+        None)
+
+  def testAgainstFile(self):
+    fio_benchmark.GenerateJobFileString(
```
Same here.

noahl replied:
Done.

yuyantingzero commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -54,6 +136,130 @@
+        name=scenario['name'],
+        rwkind=scenario['rwkind'],
+        blocksize=scenario['blocksize'],
+        iodepths=io_depths,
```
When io_depths is a list, the generated job file will be iodepths=[1,2,3,4]? Correct me if I am wrong about how jinja2 works

noahl replied:
I don't use the iodepths variable directly, there's a for loop in the template. So iodepths will be [1,2,3,4], and then that will produce four jobs in the job file, each with one of those numbers as its iodepth.

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -54,6 +136,130 @@
+
+  file_string = str(globals_template.render(filename=filename))
+  size_string = str(working_set_size) + 'G' if working_set_size else '100%'
+  for scenario in scenarios:
```
Could you move all this into the template? Mixing logic between this function and the template seems confusing.
At least prefer building up a list and calling `str.join(iterable)` to repeatedly concatenating.

cmccoy replied:
As does having multiple templates which are combined to create a single job file.

noahl replied:
I've moved it all into the template except for the filename and the size string logic. I think those would clutter the template too much, but I'm happy to change it if you still think they should be inside.

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -54,6 +136,124 @@
+verify_fatal=0
+randrepeat=0
+
+{% for scenario in scenarios %}
```
Nice. Thanks for doing this.

cmccoy commented on tests/benchmarks/fio_benchmark_test.py:
```diff
@@ -0,0 +1,119 @@
+
+
+
+[sequential_read-io-depth-1]
```
Nit: consider adding a test with multiple scenarios.

noahl replied:
Done.

noahl commented:
@yuyantingzero @cmccoy Could you take a look when you have a moment?

yuyantingzero commented:
LGTM

cmccoy commented on perfkitbenchmarker/benchmarks/fio_benchmark.py:
```diff
@@ -20,31 +20,113 @@
+
+def IODepthsValidator(string):
+  match = IODEPTHS_REGEXP.match(string)
+  return match and match.end() == len(string) and int(match.group(1)) > 0
```
Nit: adding `$` to the regex above will be equivalent to `match.end() == len(string)`.

noahl replied:
Done.

cmccoy commented:
LGTM. Thanks!
